rust/kernel/error.rs

```diff
@@ -104,3 +104,62 @@ impl From<AllocError> for Error {
         Error::ENOMEM
     }
 }
+
+// # Invariant: `-bindings::MAX_ERRNO` fits in an `i16`.
+crate::static_assert!(bindings::MAX_ERRNO as i64 <= -(i16::MIN as i64));
+
+/// Transform a kernel "error pointer" to a normal pointer.
+///
+/// Some kernel C API functions return an "error pointer" which optionally
+/// embeds an `errno`. Callers are supposed to check the returned pointer
+/// for errors. This function performs the check and converts the "error pointer"
+/// to a normal pointer in an idiomatic fashion.
+///
+/// # Safety
+///
+/// `ptr` must be valid or null.
+///
+/// # Examples
+///
+/// ```rust,no_run
+/// fn devm_platform_ioremap_resource(
+///     pdev: &mut PlatformDevice,
+///     index: u32,
+/// ) -> Result<*mut c_types::c_void> {
+///     // SAFETY: FFI call.
+///     unsafe {
+///         from_kernel_err_ptr(bindings::devm_platform_ioremap_resource(
+///             pdev.to_ptr(),
+///             index,
+///         ))
+///     }
+/// }
+/// ```
+// TODO: remove `dead_code` marker once an in-kernel client is available.
+#[allow(dead_code)]
+pub(crate) fn from_kernel_err_ptr<T>(ptr: *mut T) -> Result<*mut T> {
+    extern "C" {
+        #[allow(improper_ctypes)]
+        fn rust_helper_is_err(ptr: *const c_types::c_void) -> bool;
+
+        #[allow(improper_ctypes)]
+        fn rust_helper_ptr_err(ptr: *const c_types::c_void) -> c_types::c_long;
+    }
+
+    // SAFETY: any pointer can be safely cast to a `*const c_types::c_void`.
```

Comment:
I think we don't need `// SAFETY: ...` comments for safe code

Reply:
Good point. Do we need some sort of marker to explain why casts are valid / won't overflow? @ojeda suggested `// NOPANIC` for `.unwrap()`s before. Something along the lines of `// NO-OVERFLOW:` ?

@ojeda what's your opinion?

Reply:
Hmm, good point. I am not sure how noisy it will be (in particular inside `rust/kernel/`), but it does not hurt to try. After all, it is an operation that usually requires a bit of thinking, and many times there are better ways to do it. Having to write the comment may give people pause and search for alternatives.

For instance, here we could use [`cast()`](https://doc.rust-lang.org/std/primitive.pointer.html#method.cast) instead, and then one would not need to write a explicit comment.

In any case, I would call it `NOOVERFLOW` (we are following the `TODO`/`FIXME` etc. in that they do not have spaces/dashes/etc.). But it does look weird and long. Perhaps `CAST` is enough? i.e. "the justification for a cast".

However, if we go for `CAST`, should the justification cover more than just overflow? e.g. a pointer cast that changes to a completely unrelated type should be well-documented. But then again, someone using `cast()` should also justify it, and calling `cast()` should not be a way to avoid such a comment...

Reply:
Do we over-engineer on this? I know we want to have our code documented as good as possible. But we already have the restriction that raw pointer dererference is unsafe, so if anyone wants to cast pointer in a weird way, he/she must explain himself/herself at the point of the deref, if the pointer never gets deref, then we are safe.

In this case, the err pointer is actually from C side, and we pass it to a C function, so the safety actually replies on the C side, and we cannot do much about it on the Rust side. And I think this might be the typical case where we use raw pointers, so we might want to rethink about the bar of documenting these.

Reply:
> However, if we go for `CAST`, should the justification cover more than just overflow? e.g. a pointer cast that changes to a completely unrelated type should be well-documented. But then again, someone using `cast()` should also justify it, and calling `cast()` should not be a way to avoid such a comment...

I agree. From what I can tell playing with `cast()` in the Rust playground, that function still allows bad casts, i.e. between unrelated types. It will however catch `*const` to `*mut` casts (something that happened in the SPI PR). So it's only a partial solution.

Which means that any use of `as` or `cast()` still requires a `// CAST:` annotation - does that sound right?

It's a real pity we don't have a `void_cast()` function. Because casting to `void` is always ok so the function would not need an explicit comment.

Reply:
Also... are we mixing up two completely unrelated concepts here, just because C/Rust pointer uses the word "cast" for both of them?

In my mind, a cast between two numeric or primitive types is conceptually very different from a cast between two pointer types.

Reply:
So let me try to split the cases:
  - For safety, I agree there is no need. It should go in the proper place (i.e. the point of deref).
  - For integer arithmetic overflow (not casts), it would be too verbose and we anyway have `-Coverflow-checks` enabled by default.
  - For pointer casts, I agree it may be too annoying, which is why I was suggesting using `cast()` for those.
    + Nevertheless, people casting pointer types to wildly different ones should explain it anyhow (not for safety, but to understand what is going on).
    + It is anyway hard to enforce: we would need Clippy to support non-doc comments (same problem as for `SAFETY` ones), and we cannot use a simple script like for `SAFETY` in the stdlib because we need to know about the types.
  - For numeric casts "overflow" (e.g. `i32 as i16`, what I think @TheSven73 was thinking about), AFAIK there is no compiler option to check if truncation occurred because using `as` is explicitly asking for truncation and is not considered an error. However, given that in many cases we could be using `From`/`Into` and/or `TryFrom`/`TryInto`, the idea is that we would hope the rest of the cases would be few enough to make the comment worth it, explaining whether truncation was intended or not.
    + Or, better, we provide explicit truncation and non-truncation functions for numeric types, and then ask people to use those, instead of using `as` + a comment. We can then have a Clippy lint in this case, because we can flag any `as` usage for numeric casts.